### PR TITLE
Chimera Extension

### PR DESCRIPTION
**Pull request by jsmz97**
This PR introduces Chimera, a general-purpose achievement scalarizing function for multi-target optimization that allows users to establish a hierarchy of targets with relative or absolute thresholds for concurrent optimization.

This implementation includes a new objective class, `ChimeraObjective`, which follows a similar approach to the `DesirabilityObjective`. It scalarizes multiple targets into a single score, termed `Chimera Merits`, which is to be minimized.

For further details, please refer to the following publication:
F. Häse, L.M. Roch, and A. Aspuru-Guzik. [Chimera: enabling hierarchy-based multi-objective optimization for self-driving laboratories](https://pubs.rsc.org/ko/content/articlelanding/2018/sc/c8sc02239a#!divAbstract). *Chemical Science* **2018**, 9(39), 7642-7655.

---
### WIP:
- Further testing and validation of the `ChimeraObjective` implementation.
- Consideration of whether the optimization should be adjusted to a maximization approach, similar to the `DesirabilityObjective`.
- Additional documentation and examples to illustrate the usage of `ChimeraObjective`.

### CODE REVIEW

**Comment by Scienfitz** [comment]
@AVHopp @AdrianSosic PR in draft mode, NOT for review

File: baybe/objectives/chimera.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    """An optional tuple of target threshold types."""
+
+    softness: float = field(
+        converter=float,
+        validator=gt(0.0),
+    )
+    """The softness parameter regulating the Heaviside function."""
+
+    @targets_threshold_values.default
+    def _default_targets_threshold_values(self) -> tuple[float, ...]:
```
Comment (Scienfitz) [inline baybe/objectives/chimera.py:78]: i think there are no rasonable defaults so these should not have defaults
apart from that the defaults you specified do not satisfy the requirements `ge(0.0)`

File: baybe/objectives/chimera.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+        default_values = (ThresholdType.FRACTION,) * len(self._targets)
+        warnings.warn(
+            f"The types for target thresholds have not been specified. "
+            f"Setting the target threshold types to {default_values}.",
+            UserWarning,
+        )
+        return default_values
+
+    @softness.default  # TODO: do we need to add warning here?
+    def _default_softness(self) -> float:
```
Comment (Scienfitz) [inline baybe/objectives/chimera.py:98]: this can simply be done by writing `field(..., default=1e-3, ...)` above where the attribute is specified
no warning needed

File: baybe/objectives/chimera.py [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+            raise TypeError(
+                f"'{self.__class__.__name__}' currently only supports targets "
+                f"of type '{NumericalTarget.__name__}'."
+            )
+        if len({t.name for t in targets}) != len(targets):
+            raise ValueError("All target names must be unique.")
+        if not all(target._is_transform_normalized for target in targets):
+            raise ValueError(
+                "All targets must have normalized computational representations to "
+                "enable the computation of desirability values. This requires having "
```
Comment (Scienfitz) [inline baybe/objectives/chimera.py:114]: update message, still mentions desirability

File: baybe/objectives/chimera.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+            self.targets, transformed_threshold_values, self.targets_threshold_types
+        ):
+            if threshold_type == ThresholdType.FRACTION:
+                _threshold = threshold_value
+            elif threshold_type == ThresholdType.PERCENTILE:
+                _threshold = transformed[target.name].quantile(
+                    threshold_value, interpolation="linear"
+                )
+            elif threshold_type == ThresholdType.ABSOLUTE:
+                _threshold = threshold_value
```
[the remark this answers is not in the document]
Comment (jsmz97) [reply baybe/objectives/chimera.py:187]: All absolute threshold values are rescaled to [0, 1] via target transformation, therefore there is no difference between fractional or absolute value. Need to refactor code for clarity.

File: tests/hypothesis_strategies/objectives.py [hunk 5]
```diff
@@ -32,3 +35,75 @@ def desirability_objectives(draw: st.DrawFn):
     )
     scalarizer = draw(st.sampled_from(Scalarizer))
     return DesirabilityObjective(targets, weights, scalarizer)
+
+
+@st.composite
+def chimera_objectives(draw: st.DrawFn):
+    """Generate :class:`baybe.objectives.chimera.ChimeraObjective` and its reference."""
```
Comment (Scienfitz) [inline tests/hypothesis_strategies/objectives.py:42]: "and its reference" is not accurate anymore

File: tests/hypothesis_strategies/objectives.py [hunk 6]
```diff
@@ -32,3 +35,75 @@ def desirability_objectives(draw: st.DrawFn):
...
+    intervals = st_intervals(exclude_fully_unbounded=True, exclude_half_bounded=True)
+    targets = draw(
+        st.lists(
+            numerical_targets(intervals),
+            min_size=3,  # At least 2 targets.
```
Comment (Scienfitz) [inline tests/hypothesis_strategies/objectives.py:47]: why is it set to 3?

File: tests/hypothesis_strategies/objectives.py [hunk 7]
```diff
@@ -32,3 +35,75 @@ def desirability_objectives(draw: st.DrawFn):
...
+            # st.sampled_from(ThresholdType), # TODO: add back if the test works
+            min_size=n_targets,
+            max_size=n_targets,
+        )
+    )
+
+    # Now draw threshold values one-by-one based on each threshold type.
+    threshold_values = []
+    for i, tt in enumerate(threshold_types):
+        if tt == ThresholdType.ABSOLUTE:
```
Comment (Scienfitz) [inline tests/hypothesis_strategies/objectives.py:71]: for singletons, always use `is` and never ==

File: tests/test_objective.py [hunk 8]
```diff
@@ -93,3 +101,144 @@ def test_desirability_scalarization(values, scalarizer, weights, expected):
...
+    int_merits = int_chimera.transform(target_vals).values.flatten()
+
+    are_close = np.allclose(ext_merits, int_merits)
+    if verbose:
+        print("Chimera merits:", ext_merits)
+        print("BayBE Chimera merits:", int_merits)
+    return are_close
+
+
+@settings(max_examples=2)
```
Comment (Scienfitz) [inline tests/test_objective.py:204]: i assume this is for your testing purposes and not supposed to be here genereally?

File: tests/test_objective.py [hunk 9]
```diff
@@ -93,3 +101,144 @@ def test_desirability_scalarization(values, scalarizer, weights, expected):
...
+@given(chimera_obj=chimera_objectives(), data=st.data())
+def test_chimera_merits(chimera_obj, data):
+    """Validating chimerra merits value with external reference."""
```
Comment (Scienfitz) [inline tests/test_objective.py:207]: typo

File: tests/test_objective.py [hunk 10]
```diff
@@ -93,3 +101,144 @@ def test_desirability_scalarization(values, scalarizer, weights, expected):
...
+    # Initialize external Chimera using data from chimera_obj.
+    goals = [t.mode.value.lower() for t in chimera_obj.targets]
+    ext_chimera = Chimera(
+        tolerances=np.array(chimera_obj.threshold_values),
+        absolutes=[
+            tt.value.upper() == "ABSOLUTE" for tt in chimera_obj.threshold_types
+        ],
+        goals=[g.lower() for g in goals],
```
Comment (Scienfitz) [inline tests/test_objective.py:215]: duplicated application of lower?

File: tests/test_objective.py [hunk 11]
```diff
@@ -93,3 +101,144 @@ def test_desirability_scalarization(values, scalarizer, weights, expected):
...
+                allow_infinity=False,
+                exclude_min=True,
+                exclude_max=True,
+            ),
+        )
+        for i, t in enumerate(chimera_obj.targets)
+    ]
+
+    target_vals = data.draw(data_frames(columns=columns))
+    assume(target_vals.shape[0] >= 6)
```
Comment (Scienfitz) [inline tests/test_objective.py:240]: does `data_frames` not have options to specify the amount of rows?

Also, why 6 specifically?

**Review by Scienfitz — CHANGES_REQUESTED** [review]

File: baybe/objectives/chimera.py [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+    _threshold_values_normalized: tuple[float, ...] | None = field(
+        init=False, default=None
+    )
+    _threshold_values_shifted: tuple[float, ...] | None = field(
+        init=False, default=None
+    )
+
+    @threshold_values.default
+    def _default_threshold_values(self) -> tuple[float, ...]:
+        default_values = (0.0,) * len(self._targets)
```
Comment (Scienfitz) [inline baybe/objectives/chimera.py:96]: 0.0 doesnt seem like a reasonable default to me, whats the recommended default by Chimera?

File: baybe/objectives/chimera.py [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+                "All targets must have normalized computational representations to "
+                "enable the computation of chimera merit values. This requires having "
+                "appropriate target bounds and transformations in place."
+            )
+
+    @threshold_values.validator
+    def _validate_threshold_values(self, _, values) -> None:
+        if (lv := len(values)) != (lt := len(self._targets)):
+            raise ValueError(
+                f"If custom threshold values are specified, there must be one for each target. "  # noqa: E501
```
Comment (Scienfitz) [inline baybe/objectives/chimera.py:134]: why the noqa? you can break the strings anywhere

File: baybe/objectives/chimera.py [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+    def _validate_threshold_types(self, _, types) -> None:
+        if (lt := len(types)) != (ltg := len(self._targets)):
+            raise ValueError(
+                f"If custom threshold types are specified, there must be one for each target. "  # noqa: E501
+                f"Specified number of targets: {ltg}. Specified number of threshold types: {lt}."  # noqa: E501
+            )
+
+    def _soft_heaviside(self, value: float, softness: float) -> float:
+        arg = -value / softness
+        return np.exp(-np.logaddexp(0, arg))
```
Comment (Scienfitz) [inline baybe/objectives/chimera.py:148]: is there any reason for suing this strange expression? 
isnt this equal to `1.0 / (1.0 + np.exp(arg))`?

File: baybe/objectives/chimera.py [hunk 15]
```diff
@@ -0,0 +1,395 @@
...
+            float
+        )  # Pandas handles booleans as floats automatically
+
+    def step(self, value: float, softness: float = 1e-6) -> float:
+        """Apply a step function to the given value based on the specified softness.
+
+        Args:
+            value: The input value to apply the step function to.
+            softness: The softness parameter for the step function.
+                If less than 1e-5, a hard Heaviside step function is used.
```
Comment (Scienfitz) [inline baybe/objectives/chimera.py:161]: why is this kind of distinction even done?

File: baybe/objectives/chimera.py [hunk 16]
```diff
@@ -0,0 +1,395 @@
...
+                    "to `True` when left unspecified. However, this behavior will be "
+                    "changed in a future version. If you want to invoke the old "
+                    "behavior, please explicitly set `allow_extra=True`.",
+                    DeprecationWarning,
+                )
+        # <<<<<<<<<< Deprecation
+
+        # Extract the relevant part of the dataframe
+        # targets = get_chimera_transform_objects(
+        #     df, self.targets, allow_missing=allow_missing, allow_extra=allow_extra
```
Comment (Scienfitz) [inline baybe/objectives/chimera.py:314]: why is this here

File: baybe/objectives/chimera.py [hunk 17]
```diff
@@ -0,0 +1,395 @@
...
+        # )
+        targets = get_transform_objects(
+            df, self.targets, allow_missing=allow_missing, allow_extra=allow_extra
+        )
+        transformed = df[[t.name for t in targets]].copy()
+        _threshold_values_transformed = list(self.threshold_values)
+        # Helper function to transform the target-specific threshold value
```
Comment (Scienfitz) [inline baybe/objectives/chimera.py:321]: there is no function here

**Review by Scienfitz — CHANGES_REQUESTED** [review]

File: baybe/objectives/chimera.py [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+    PERCENTILE = "PERCENTILE"
+    """The target threshold is a percentile value."""
+
+    FRACTION = "FRACTION"
+    """The target threshold is a fraction value."""
+
+
+@define(frozen=True, slots=False)
+class ChimeraObjective(Objective):
+    """An objective scalarizing multiple targets using desirability values."""
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply baybe/objectives/chimera.py]: Corrected in the latest version

File: baybe/objectives/chimera.py [hunk 19]
```diff
@@ -0,0 +1,362 @@
...
+
+    _targets: tuple[Target, ...] = field(
+        converter=to_tuple,
+        validator=[min_len(2), deep_iterable(member_validator=instance_of(Target))],
+        alias="targets",
+    )
+    "The targets considered by the objective."
+
+    targets_threshold_values: tuple[float, ...] = field(
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply baybe/objectives/chimera.py]: Removed `targets_` prefix in naming in the latest version

File: baybe/objectives/chimera.py (hunk 1 above)
Comment (jsmzhng) [reply baybe/objectives/chimera.py]: Removed

File: baybe/objectives/chimera.py [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+            raise ValueError(
+                f"If custom threshold types are specified, there must be one for each target. "  # noqa: E501
+                f"Specified number of targets: {ltg}. Specified number of threshold types: {lt}."  # noqa: E501
+            )
+
+    def _soft_heaviside(self, value: float, softness: float) -> float:
+        arg = -value / softness
+        return np.exp(-np.logaddexp(0, arg))
+
+    def _hard_heaviside(self, value: float) -> float:
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply baybe/objectives/chimera.py:146]: It is still needed to ensure the consistency with the reference implementation and the tests.

File: tests/hypothesis_strategies/objectives.py (hunk 6 above)
Comment (jsmzhng) [reply tests/hypothesis_strategies/objectives.py]: - three intended for initial testing
- Already changed the condition to 2

File: tests/hypothesis_strategies/objectives.py (hunk 7 above)
Comment (jsmzhng) [reply tests/hypothesis_strategies/objectives.py]: Corrected

File: tests/test_objective.py (hunk 8 above)
Comment (jsmzhng) [reply tests/test_objective.py]: - max_examples set to 100 instead
- Fraction case working, absolute case still missing (percentile case no comparison)

File: tests/test_objective.py (hunk 9 above)
Comment (jsmzhng) [reply tests/test_objective.py:272]: Corrected

File: tests/test_objective.py (hunk 10 above)
Comment (jsmzhng) [reply tests/test_objective.py]: refactored for simplicity

File: tests/hypothesis_strategies/objectives.py [hunk 21]
```diff
@@ -32,3 +35,75 @@ def desirability_objectives(draw: st.DrawFn):
...
+
+@st.composite
+def chimera_objectives(draw: st.DrawFn):
+    """Generate :class:`baybe.objectives.chimera.ChimeraObjective` and its reference."""
+    intervals = st_intervals(exclude_fully_unbounded=True, exclude_half_bounded=True)
+    targets = draw(
+        st.lists(
+            numerical_targets(intervals),
+            min_size=3,  # At least 2 targets.
+            max_size=4,  # TODO: remove max_size once tests working
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/hypothesis_strategies/objectives.py]: - condition loosed in newer updates
- corrections for ABSOLUTE case pending

File: baybe/objectives/chimera.py (hunk 15 above)
Comment (jsmzhng) [reply baybe/objectives/chimera.py:157]: - this is here to ensure comparability with the reference implementation

File: baybe/objectives/chimera.py (hunk 17 above)
Comment (jsmzhng) [reply baybe/objectives/chimera.py]: Already refactored in the newer version (lines obselete)

File: baybe/objectives/chimera.py (hunk 16 above)
Comment (jsmzhng) [reply baybe/objectives/chimera.py]: - removed in newer version, old idea to initialize targets but no longer needed

File: baybe/objectives/chimera.py (hunk 14 above)
Comment (jsmzhng) [reply baybe/objectives/chimera.py:144]: - Yes both expressions are identical
- The exp(-logaddexp(…)) form is more numerically stable:`np.logaddexp(0, arg)` computes `log(1 + exp(arg))` in a way that avoids overflow when arg is large `(> ≈700)` and `under-flow` when it is very negative.

File: baybe/objectives/chimera.py (hunk 13 above)
Comment (jsmzhng) [reply baybe/objectives/chimera.py]: Modified in the newer version

File: tests/hypothesis_strategies/objectives.py [hunk 22]
```diff
@@ -32,3 +35,75 @@ def desirability_objectives(draw: st.DrawFn):
...
+            ),  # TODO: let's test these two first
+            # st.sampled_from(ThresholdType), # TODO: add back if the test works
+            min_size=n_targets,
+            max_size=n_targets,
+        )
+    )
+
+    # Now draw threshold values one-by-one based on each threshold type.
+    threshold_values = []
+    for i, tt in enumerate(threshold_types):
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/hypothesis_strategies/objectives.py:76]: Changed in the newer version

File: tests/hypothesis_strategies/objectives.py [hunk 23]
```diff
@@ -32,3 +35,75 @@ def desirability_objectives(draw: st.DrawFn):
...
+        st.lists(
+            numerical_targets(intervals),
+            min_size=3,  # At least 2 targets.
+            max_size=4,  # TODO: remove max_size once tests working
+            unique_by=lambda t: t.name,
+        )
+    )
+
+    # Force that every drawn target has a linear transformation and mode MIN or MAX.
+    assume(all(t.mode is not TargetMode.MATCH for t in targets))
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/hypothesis_strategies/objectives.py]: because match is not implemented in the reference code, nothing to compare and to test here

File: tests/hypothesis_strategies/objectives.py (hunk 6 above)
Comment (Scienfitz) [reply tests/hypothesis_strategies/objectives.py]: commend above not accurate in this case

File: tests/test_objective.py (hunk 8 above)
Comment (Scienfitz) [reply tests/test_objective.py]: why are you setting it at all (other hypothesis tests dont set it)?

File: tests/hypothesis_strategies/objectives.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply tests/hypothesis_strategies/objectives.py]: if you apply the target transformations before passing them to the external chimera step (as discussed many times) this restriction should not be needed anymore (at the transformed match target is maximizing then)

File: baybe/objectives/chimera.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/objectives/chimera.py:146]: discussed

File: tests/test_objective.py (hunk 8 above)
Comment (jsmzhng) [reply tests/test_objective.py]: It is set globally, remove max_examples later when implementation ready

File: tests/test_objective.py [hunk 24]
```diff
@@ -243,26 +245,27 @@ def compare_merits(
...
-
-    if int_const or ext_degenerate:
-        return True
-
+    # Normalize external merits to match internal implementation
+    ext_range = np.amax(ext_merits) - np.amin(ext_merits)
+    if ext_range > 0:
+        ext_merits = (ext_merits - np.amin(ext_merits)) / ext_range
+    else:
+        ext_merits = np.zeros_like(ext_merits)  # Handle uniform values
```
Comment (jsmzhng) [inline tests/test_objective.py:261]: @Scienfitz The reference code performs final normalization in `scalarize` via:
```
if np.amax(merits) > 0.:
    merits = (merits - np.amin(merits)) / (np.amax(merits) - np.amin(merits))
```
See reference: [reference](https://github.com/aspuru-guzik-group/chimera/blob/master/src/chimera/chimera.py) line 259 onwards

This approach has two issues:
- It is not numerically stable: Division by (max - min) can cause division-by-zero when values are nearly identical
- When max(merits) ≤ 0, no normalization occurs, leaving merits in arbitrary negative ranges, causing confusions as they are acquisition weights for BoTorch (in minimization mode), our implementation always normalizes to [0,1] when range > 0 and handling edge case of a uniform merits

Comment (jsmzhng) [reply tests/test_objective.py:261]: @Scienfitz this however does not mitigate the warning message when running the test:
```
tests/test_objective.py::test_chimera_merits
  /Users/x272626/Desktop/chimera/src/chimera/chimera.py:263: RuntimeWarning: invalid value encountered in divide
    merits = (merits - np.amin(merits)) / (np.amax(merits) - np.amin(merits))
 ```
 which means we have uniform merits (all values identical) -> this is however not sth that we can explicitly predict based on the input values, nor build the exclusion logics explicitly in the test as many different combinations and variations in `target_vals` and `threshold_vals` and `threshold_type` can lead to this issue.
 
How can we overcome this?

File: tests/test_objective.py [hunk 25]
```diff
@@ -93,3 +101,191 @@ def test_desirability_scalarization(values, scalarizer, weights, expected):
...
+class TestInvalidChimeraObjectiveCreation:
+    """Tests for invalid ChimeraObjective creation."""
+
+    two_targets = [
+        NumericalTarget(name="Target_1", mode="MIN", bounds=(0, 100)),
+        NumericalTarget(name="Target_2", mode="MIN", bounds=(0, 100)),
+    ]
+
+    def test_empty_target_list(self):
+        with pytest.raises(ValueError):
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/test_objective.py:121]: Use `@property` to avoid creating `NumericalTarget` objects at import time, which prevents transformation warnings during file loading

[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/test_objective.py:121]: `match` is an allowed argument hence no need to exclude them?

File: tests/test_objective.py (hunk 8 above)
Comment (Scienfitz) [reply tests/test_objective.py]: you need to remove it already for testing as otherwise only 2 samples will be tested and coverage will be bad

File: tests/test_objective.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply tests/test_objective.py:121]: dont know how your comments relate to this thread

what I mean:
```python
with pytest.raises(ValueError, match="some substring of the expected error message"):
```

File: tests/test_objective.py (hunk 8 above)
Comment (jsmzhng) [reply tests/test_objective.py]: Already noticed and removed

File: tests/test_objective.py [hunk 26]
```diff
@@ -93,3 +104,204 @@ def test_desirability_scalarization(values, scalarizer, weights, expected):
...
+                exclude_min=True,
+                exclude_max=True,
+            ),
+        )
+        for target in chimera_obj.targets
+    ]
+
+    target_vals = data.draw(
+        data_frames(columns=columns, index=range_indexes(min_size=2)).filter(
+            lambda df: len(df.drop_duplicates()) > 1
```
Comment (jsmzhng) [inline tests/test_objective.py:289]: Is this necessary? Because it could have multiple measurements that are the same

File: tests/hypothesis_strategies/objectives.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/hypothesis_strategies/objectives.py]: Match target already reactivated

File: tests/hypothesis_strategies/objectives.py [hunk 27]
```diff
@@ -32,3 +35,75 @@ def desirability_objectives(draw: st.DrawFn):
...
+    n_targets = len(targets)
+
+    threshold_types = draw(
+        st.lists(
+            st.sampled_from(
+                [ThresholdType.FRACTION, ThresholdType.PERCENTILE]
+            ),  # TODO: let's test these two first
+            # st.sampled_from(ThresholdType), # TODO: add back if the test works
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/hypothesis_strategies/objectives.py]: This is already enabled

File: tests/hypothesis_strategies/objectives.py [hunk 28]
```diff
@@ -32,3 +35,75 @@ def desirability_objectives(draw: st.DrawFn):
...
+                    exclude_min=True,
+                    exclude_max=True,
+                    allow_nan=False,
+                    allow_infinity=False,
+                )
+            )
+        threshold_values.append(val)
+
+    # softness = draw(st.floats(min_value=0.0, max_value=1.0))
+    softness = 0.0  # TODO: remove this line once tests working
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/hypothesis_strategies/objectives.py]: Softness has to be tested as well, to add back the commented line

File: tests/hypothesis_strategies/objectives.py [hunk 29]
```diff
@@ -32,3 +35,75 @@ def desirability_objectives(draw: st.DrawFn):
...
+
+    # Initialize internal ChimeraObjective using the drawn targets.
+    int_chimera = ChimeraObjective(
+        targets=targets,
+        threshold_values=tuple(threshold_values),
+        threshold_types=tuple(tt.value for tt in threshold_types),
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/hypothesis_strategies/objectives.py:103]: No need to `.value`, just `threshold_types = threshold_types` should be fine

File: tests/test_objective.py [hunk 30]
```diff
@@ -1,14 +1,22 @@
 """Tests for the objective module."""
 
+import hypothesis.strategies as st
 import numpy as np
+import pandas as pd
 import pytest
 from cattrs import IterableValidationError
+from chimera import Chimera
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/test_objective.py:8]:
```
test = [
    "hypothesis[pandas]>=6.88.4",
    "tenacity>=8.5.0",
    "pytest>=7.2.0",
    "pytest-cov>=4.1.0",
]
```
add Chimera dependency in `pyproject.toml`

File: tests/test_objective.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/test_objective.py:121]: Error originated from `baybe/objectives/chimera.py`

```
    _targets: tuple[Target, ...] = field(
        converter=to_tuple,
        validator=[min_len(2), deep_iterable(member_validator=instance_of(Target))],
        alias="targets",
    )
```

File: tests/test_objective.py [hunk 31]
```diff
@@ -93,3 +101,191 @@ def test_desirability_scalarization(values, scalarizer, weights, expected):
...
+        col_tr = 1.0 - target.transform(target_vals[target.name])
+        target_vals_transformed[target.name] = col_tr
+
+        if ttype is ThresholdType.ABSOLUTE:
+            # transform + invert, but **do not** min-max scale here
+            thr_tr = 1.0 - target.transform(pd.Series([tval])).values[0]
+            _threshold_values_transformed[int_chimera.targets.index(target)] = thr_tr
+            ext_absolutes.append(
+                True
+            )  # TODO: this is wrong, no append, always indexing
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/test_objective.py]: Already resolved

File: tests/test_objective.py [hunk 32]
```diff
@@ -93,3 +101,191 @@ def test_desirability_scalarization(values, scalarizer, weights, expected):
...
+    """Validating chimerra merits value with external reference."""
+    # Initialize external Chimera using data from chimera_obj.
+    goals = [t.mode.value.lower() for t in chimera_obj.targets]
+    threshold_vals = np.array(chimera_obj.threshold_values)
+    ext_chimera = Chimera(
+        tolerances=threshold_vals,
+        absolutes=[
+            True
+            for tt in chimera_obj.threshold_types
+            # TODO: because of the transform, all become absolute
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply tests/test_objective.py]: Already updated to correct version

File: baybe/objectives/chimera.py [hunk 33]
```diff
@@ -0,0 +1,395 @@
...
+    _targets: tuple[Target, ...] = field(
+        converter=to_tuple,
+        validator=[min_len(2), deep_iterable(member_validator=instance_of(Target))],
+        alias="targets",
+    )
+    "The targets considered by the objective."
+
+    threshold_values: tuple[float, ...] = field(
+        converter=lambda w: cattrs.structure(w, tuple[float, ...]),
+        validator=deep_iterable(member_validator=[finite_float, ge(0.0)]),
```
[the remark this answers is not in the document]
Comment (jsmzhng) [reply baybe/objectives/chimera.py]: Already corrected (negative threshold allowed in case of absolute threshold for e.g. negative ranged values)

File: baybe/objectives/chimera.py (hunk 12 above)
Comment (jsmzhng) [reply baybe/objectives/chimera.py]: default threshold values are completely removed
